core/src/main/java/org/apache/druid/data/input/ObjectSource.java

```diff
@@ -0,0 +1,116 @@
+ */
+
+package org.apache.druid.data.input;
+
+import com.google.common.base.Predicate;
+import org.apache.druid.guice.annotations.ExtensionPoint;
+import org.apache.druid.java.util.common.FileUtils;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.logger.Logger;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * ObjectSource abstracts an object and knows how to read bytes from the given object.
+ */
+@ExtensionPoint
+public interface ObjectSource<T>
+{
+  Logger LOG = new Logger(ObjectSource.class);
+
+  int DEFAULT_FETCH_BUFFER_SIZE = 4 * 1024; // 4 KB
+  int DEFAULT_MAX_FETCH_RETRY = 2; // 3 tries including the initial try
+
+  /**
+   * CleanableFile is the result type of {@link #fetch}.
+   * It should clean up any temporary resource on {@link #close()}.
+   */
+  interface CleanableFile extends Closeable
+  {
+    File file();
+  }
+
+  T getObject();
+
+  /**
+   * Opens an {@link InputStream} on the object directly.
+   * This is the basic way to read the given object.
+   *
+   * @see #fetch as an alternative way to read data.
+   */
+  InputStream open() throws IOException;
+
+  /**
+   * Fetches the object into the local storage.
+   * This method might be preferred instead of {@link #open()}, for example
+   *
+   * - {@link org.apache.druid.data.input.impl.InputFormat} requires expensive random access on remote storage.
+   * - Holding a connection until you consume the entire InputStream is expensive.
+   *
+   * @param temporaryDirectory to store temp data. This directory will be removed automatically once
+   *                           the task finishes.
+   * @param fetchBuffer        is used to fetch remote object into local storage.
+   *
+   * @see FileUtils#copyLarge
+   */
+  default CleanableFile fetch(File temporaryDirectory, byte[] fetchBuffer) throws IOException
+  {
+    final File tempFile = File.createTempFile("druid-object-source", ".tmp", temporaryDirectory);
+    LOG.debug("Fetching object into file[%s]", tempFile.getAbsolutePath());
+    try (InputStream is = open()) {
+      FileUtils.copyLarge(
+          is,
+          tempFile,
+          fetchBuffer,
+          getRetryCondition(),
+          DEFAULT_MAX_FETCH_RETRY,
+          StringUtils.format("Failed to fetch into [%s]", tempFile.getAbsolutePath())
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return new CleanableFile()
+    {
+      @Override
+      public File file()
+      {
+        return tempFile;
+      }
+
+      @Override
+      public void close()
+      {
+        if (!tempFile.delete()) {
+          LOG.warn("Failed to remove file[%s]", tempFile.getAbsolutePath());
+        }
+      }
+    };
+  }
+
+  /**
+   * {@link #fetch} will retry during the fetch if it sees an exception mathing to the returned predicate.
+   */
+  Predicate<Throwable> getRetryCondition();
```

Comment:
Suggest calling this `getFetchRetryCondition()` since it's only used for fetch

Reply:
👍